### PR TITLE
Adding JLBP-17: Coordinate Rollout of Breaking Changes

### PR DESCRIPTION
**Pull request by garrettjonesgoogle**

### CODE REVIEW

File: library-best-practices/JLBP-17.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+[JLBP-17] Coordinate Major Version Adoption
+-------------------------------------------
+
+For any particular library dependency chain, the libraries involved need to
+coordinate the adoption of a new major version of any particular shared library
+which doesn't rename its package so that consumers can continue using all
+functionality together. (This best practice doesn't apply to adoption of new
+major versions where the java package was renamed, because the old and new
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: delete "java"

File: library-best-practices/JLBP-17.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
+[JLBP-17] Coordinate Major Version Adoption
+-------------------------------------------
+
+For any particular library dependency chain, the libraries involved need to
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: This sentence is long and a tad hard to follow. Can we shorten and break it up?

File: library-best-practices/JLBP-17.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+dependency chain (both gax-java and grpc-java depend on guava 20.0):
+
+```
+gax-java 1.29.0 ->
+  grpc-java 1.13.1 ->
+    guava 20.0
+  guava 20.0
+```
+
+Next, guava releases a later version (26.0) which has a breaking change (a
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: Suppose Guava releases version 26.0 which deletes a public method.

File: library-best-practices/JLBP-17.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+method is deleted). If gax-java 1.29.0 and grpc-java 1.13.1 depend on the
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: "If gax-java 1.29.0 and grpc-java 1.13.1 depend on the
feature that is deleted, then it would be impossible" --> "If gax-java 1.29.0 and grpc-java 1.13.1 invoke that method, it is impossible"

File: library-best-practices/JLBP-17.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+feature that is deleted, then it would be impossible for a user to depend on the
+latest version of all of these artifacts. The following dependency chain will
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: all of these artifacts --> all three artifacts

File: library-best-practices/JLBP-17.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+have static linkage conflicts (this assumes that the end user overrides
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: will have --> has

In general, present tense. Also rewrite without parentheticals.

File: library-best-practices/JLBP-17.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+```
+gax-java 1.29.0 ->
+  grpc-java 1.13.1 ->
+    (ERROR) guava 26.0
+  (ERROR) guava 26.0
+```
+
+After grpc-java releases a new version (1.16.0) which depends on the new version
+of guava, gax-java will still experience conflicts:
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: will still experience --> still experiences

File: library-best-practices/JLBP-17.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+```
+gax-java 1.29.0 ->
+  grpc-java 1.16.0 ->
+    guava 26.0
+  (ERROR) guava 26.0
+```
+
+Finally after gax-java upgrades its dependency on Guava, there will no longer be
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: Only when gax-java and grpc-java both upgrade their dependency on Guava are there no longer any conflicts:

File: library-best-practices/JLBP-17.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+gax-java 1.35.0 ->
+  grpc-java 1.16.0 ->
+    guava 26.0
+  guava 26.0
+```
+
+This state is impossible to avoid completely (particularly during the process of
+rolling out such a breaking change), but the amount of time that the dependency
+chain is in this state should be minimized. Each library in the chain should be
+prepared for such an upgrade before it starts, and the upgrade should be
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: starts, and the  --> "starts; and the" or just break into two sentences

File: library-best-practices/JLBP-17.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+propagated up the tree in as little time as possible, so that consumers can
+continue to safely use the rule of depending on the latest version of every
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: use the rule of depending --> depend

File: library-best-practices/JLBP-17.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+package in the tree.
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: delete "in the tree"

File: library-best-practices/JLBP-17.md (hunk 1 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Will do.

File: library-best-practices/JLBP-17.md (hunk 2 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: I will rewrite the paragraph.

File: library-best-practices/JLBP-17.md (hunk 3 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Will do.

File: library-best-practices/JLBP-17.md (hunk 4 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Will do.

File: library-best-practices/JLBP-17.md (hunk 5 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Will do.

File: library-best-practices/JLBP-17.md (hunk 6 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: I still have to internalize this present tense thing :-P

File: library-best-practices/JLBP-17.md (hunk 7 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: ack'n'd

File: library-best-practices/JLBP-17.md (hunk 8 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Sounds good.

File: library-best-practices/JLBP-17.md (hunk 9 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: I'll do that.

File: library-best-practices/JLBP-17.md (hunk 10 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Nice.

File: library-best-practices/JLBP-17.md (hunk 11 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Sounds good.

**Comment by garrettjonesgoogle** [comment]
PR feedback addressed, PTAL

File: library-best-practices/JLBP-17.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+-------------------------------------------
+
+For any particular library dependency chain, the libraries involved need to
+coordinate the adoption of new major versions of any shared dependency that has
+breaking changes. This is necessary so that consumers can easily use the latest
+version of all of the functionality together.  This best practice doesn't apply
+to the adoption of new major versions where the package was renamed, because the
+old and new packages can co-exist.
+
+As an example of what happens when this is not followed, imagine the partial
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: we can drop "imagine the partial
adoption of a new major version of a leaf library." since you cover this and two other cases in sequence here.

File: library-best-practices/JLBP-17.md [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+```
+gax-java 1.29.0 ->
+  grpc-java 1.13.1 ->
+    guava 20.0
+  guava 20.0
+```
+
+Suppose Guava version 26.0 deletes a public method. If gax-java 1.29.0 and
+grpc-java 1.13.1 invoke that method, it is impossible for a user to depend on
+the latest version of all three artifacts because they would experience static
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: "because they would experience" --> without

File: library-best-practices/JLBP-17.md (hunk 12 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: I think it helps to know what's coming though, so you know why it's relevant information.

File: library-best-practices/JLBP-17.md (hunk 13 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Ok that seems fine.

**Comment by garrettjonesgoogle** [comment]
I have addressed more feedback, PTAL

File: library-best-practices/JLBP-17.md [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+For any particular library dependency chain, the libraries involved need to
+coordinate the adoption of new major versions of any shared dependency that has
+breaking changes. This is necessary so that consumers can easily use the latest
+version of all of the functionality together.  This best practice doesn't apply
+to the adoption of new major versions where the package was renamed, because the
+old and new packages can co-exist.
+
+Below is an example of what happens when a new major version of a leaf library
+is partially adopted. Start with the following dependency chain (both gax-java
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: "Start with the following dependency chain (both " --> Suppose

File: library-best-practices/JLBP-17.md [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+```
+gax-java 1.29.0 ->
+  grpc-java 1.13.1 ->
+    guava 20.0
+  guava 20.0
+```
+
+Suppose Guava version 26.0 deletes a public method. If gax-java 1.29.0 and
+grpc-java 1.13.1 invoke that method, it is impossible for a user to depend on
+the latest version of all three artifacts without static linkage conflicts. The
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: latest version of Guava is not necessarily 26

File: library-best-practices/JLBP-17.md (hunk 2 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: yes...

File: library-best-practices/JLBP-17.md (hunk 14 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Will do

File: library-best-practices/JLBP-17.md (hunk 15 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: I will clarify that in the example, it is the latest version.

**Comment by garrettjonesgoogle** [comment]
Third round of feedback addressed, PTAL

File: library-best-practices/JLBP-17.md [hunk 16]
```diff
@@ -0,0 +1,61 @@
+[JLBP-17] Coordinate Major Version Adoption
+-------------------------------------------
+
+For any particular library dependency chain, the libraries involved need to
+coordinate the adoption of new major versions of any shared dependency that has
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: Can we shorten this to "new major versions of shared dependencies."?

File: library-best-practices/JLBP-17.md [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+breaking changes. This is necessary so that consumers can easily use the latest
+version of all of the functionality together.  This best practice doesn't apply
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: of all of --> of all

File: library-best-practices/JLBP-17.md [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+to the adoption of new major versions where the package was renamed, because the
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: delete "the adoption of"

File: library-best-practices/JLBP-17.md [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+is partially adopted. Suppose gax-java and grpc-java depend on guava 20.0:
+
+```
+gax-java 1.29.0 ->
+  grpc-java 1.13.1 ->
+    guava 20.0
+  guava 20.0
+```
+
+Suppose Guava version 26.0 deletes a public method and it is now the latest
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: So this example is suggesting that Guava should coordinate releases with gax-java and grpc-java? That criterion seems only satisfiable in a mono-repo world.

File: library-best-practices/JLBP-17.md (hunk 16 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Done.

File: library-best-practices/JLBP-17.md (hunk 17 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Done.

File: library-best-practices/JLBP-17.md (hunk 18 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Done.

File: library-best-practices/JLBP-17.md (hunk 19 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Yes, but only for major versions. No coordination needs to be done for minor (= non-breaking) changes.

**Comment by garrettjonesgoogle** [comment]
Fourth round of feedback addressed, PTAL

File: library-best-practices/JLBP-17.md [hunk 20]
```diff
@@ -0,0 +1,30 @@
+[JLBP-17] Coordinate Major Version Adoption
+-------------------------------------------
+
+It is clearly undesirable for any particular library dependency tree to contain
+conflicts. Unfortunately, for any consumer of such a dependency tree that tries
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: "Unfortunately, for any consumer of such a dependency tree that tries
to use the highest versions available of all libraries contained in it, there
are conflicts as soon as a breaking change is introduced."

I'm not sure I follow this sentence.

File: library-best-practices/JLBP-17.md [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+to use the highest versions available of all libraries contained in it, there
+are conflicts as soon as a breaking change is introduced. Thus, breaking changes
+need to be propagated as quickly as possible so that such consumers can continue
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: delete "such"

File: library-best-practices/JLBP-17.md [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+using the highest versions.
+
+Perform the rollout in this manner:
+
+1. Decide whether to break the feature as an atomic change or to perform the
+   change in two passes (deprecate then delete).
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: deprecate,

File: library-best-practices/JLBP-17.md [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+   - If the feature is used by stable code in other libraries, you must use two
+     passes (as per [JLBP-7](JLBP-7.md)).
+   - If there does not appear to be usage of the code in other libraries, an
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: If the code does not appear to be used

File: library-best-practices/JLBP-17.md [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+     in-place breakage may be ok.
+   - If the feature is used by unstable code or stable code having little usage
+     itself, prefer to use two passes if possible, with in-place breakage only
+     if two passes is judged to be too costly.
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: are judged

File: library-best-practices/JLBP-17.md [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+Perform the rollout in this manner:
+
+1. Decide whether to break the feature as an atomic change or to perform the
+   change in two passes (deprecate then delete).
+   - If the feature is used by stable code in other libraries, you must use two
+     passes (as per [JLBP-7](JLBP-7.md)).
+   - If there does not appear to be usage of the code in other libraries, an
+     in-place breakage may be ok.
+   - If the feature is used by unstable code or stable code having little usage
+     itself, prefer to use two passes if possible, with in-place breakage only
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: delete "to use"

File: library-best-practices/JLBP-17.md [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+     if two passes is judged to be too costly.
+2. Make sure that consuming libraries are prepared for the breakage.
+   - In the case of in-place breakage, have a PR prepared and approved in the
+     consuming libraries that switches from the old surface to the new surface.
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: This doesn't work. The consuming won't approve such a PR until the continuous build passes, which it won't at this stage.

File: library-best-practices/JLBP-17.md (hunk 20 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: I will rewrite it.

File: library-best-practices/JLBP-17.md (hunk 21 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: I will reword in a different way.

File: library-best-practices/JLBP-17.md (hunk 22 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: I assume comma before delete actually? (That's the change I'll try out)

File: library-best-practices/JLBP-17.md (hunk 23 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Will do.

File: library-best-practices/JLBP-17.md (hunk 24 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Will do.

File: library-best-practices/JLBP-17.md (hunk 25 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Will do.

File: library-best-practices/JLBP-17.md (hunk 26 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: I mean provisionally approved - formal approval would wait for the actual roll out. I will clarify.

**Comment by garrettjonesgoogle** [comment]
Rewrite-feedback addressed, PTAL

File: library-best-practices/JLBP-17.md [hunk 27]
```diff
@@ -0,0 +1,34 @@
+[JLBP-17] Coordinate Major Version Adoption
+-------------------------------------------
+
+It is clearly undesirable for a dependency tree to contain conflicts.
+Unfortunately, as soon as any library in a dependency tree introduces a breaking
+change, consumers can't use the latest versions of those libraries together. To
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: "those libraries" is unclear. I guess it means the libraries in the dependency tree?

File: library-best-practices/JLBP-17.md [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+support users who try to generally use the latest versions of their
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: This sentence can be cut down a bit. Maybe

To enable clients to depend on the latest versions of their dependencies,  breaking changes should be propagated as quickly as possible.

File: library-best-practices/JLBP-17.md [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+   change in two passes, deprecate then delete.
+   - If the feature is used by stable code in other libraries, you must use two
+     passes (as per [JLBP-7](JLBP-7.md)).
+   - If the code does not appear to be used in other libraries, an in-place
+     breakage may be ok.
+   - If the feature is used by unstable code or stable code having little usage
+     itself, prefer two passes if possible, with in-place breakage only if two
+     passes are judged to be too costly.
+2. Make sure that consuming libraries are prepared for the breakage.
+   - In the case of in-place breakage, have PRs prepared and provisionally
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: I'm still not comfortable with this. I would personally not approve a change that didn't pass CI, provisionally or otherwise.

File: library-best-practices/JLBP-17.md [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+     approved in the consuming libraries that switch from the old surface to the
+     new surface. These PRs of course will not pass CI, but the author should at
+     least verify that the code builds locally before requesting a review.
+   - In the case of a two-pass breakage, mark the undesired surface as
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: perform a release --> release

File: library-best-practices/JLBP-17.md [hunk 31]
```diff
@@ -0,0 +1,34 @@
...
+     `@Deprecated` and perform a release. Make sure that all consuming libraries
+     have removed their references to the deprecated functionality. This invokes
+     [JLBP-13](JLBP-13.md) for consuming libraries.
+3. Release the breakage and make sure that the version propagates up the
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: breakage --> incompatible version

File: library-best-practices/JLBP-17.md (hunk 27 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Yes. I will qualify it better.

File: library-best-practices/JLBP-17.md (hunk 28 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Sounds good.

File: library-best-practices/JLBP-17.md (hunk 29 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Well maybe "approval" is the wrong word, but the point is that the author has addressed your concerns with the code. The intent is to make the final approval quick, and not requiring a bunch of back & forth to address issues that can be caught even if CI fails. I will try to capture that somehow.

File: library-best-practices/JLBP-17.md (hunk 30 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Ok

File: library-best-practices/JLBP-17.md (hunk 31 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Will do

**Comment by garrettjonesgoogle** [comment]
Another day, another round of feedback addressed. PTAL @elharo

File: library-best-practices/JLBP-17.md [hunk 32]
```diff
@@ -0,0 +1,38 @@
+[JLBP-17] Coordinate Major Version Adoption
+-------------------------------------------
+
+It is clearly undesirable for a dependency tree to contain conflicts.
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: I think we can delete "It is clearly undesirable for a dependency tree to contain conflicts. Unfortunately,"

If it's clear, we don't need to restate it here.

File: library-best-practices/JLBP-17.md [hunk 33]
```diff
@@ -0,0 +1,38 @@
...
+Unfortunately, as soon as any library in a dependency tree introduces a breaking
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: Suggested rewording:

When a library in a dependency tree introduces a breaking
change, consumers can't update to that version until all their other dependencies which use the library also update to that version

Is that a fair restatement?

File: library-best-practices/JLBP-17.md [hunk 34]
```diff
@@ -0,0 +1,38 @@
...
+change, consumers can't use the latest versions of those libraries in the
+dependency tree together. To enable clients to depend on the latest versions of
+their dependencies, breaking changes should be propagated as quickly as
+ possible.
+
+Perform the rollout in this manner:
+
+1. Decide whether to break the feature as an atomic change or to perform the
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: Decide whether to introduce the incompatibility in a single release or to perform the change over two releases: first deprecating, then deleting.

File: library-best-practices/JLBP-17.md [hunk 35]
```diff
@@ -0,0 +1,38 @@
...
+   change in two passes, deprecate then delete.
+   - If the feature is used by stable code in other libraries, you must use two
+     passes (as per [JLBP-7](JLBP-7.md)).
+   - If the code does not appear to be used in other libraries, an in-place
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: Some breaking changes may not be able to be done with a deprecation first. For example, changing a class to final or removing or adding a checked exception in a method's throws clause. Should we mention this?

File: library-best-practices/JLBP-17.md [hunk 36]
```diff
@@ -0,0 +1,38 @@
...
+     breakage may be ok.
+   - If the feature is used by unstable code or stable code having little usage
+     itself, prefer two passes if possible, with in-place breakage only if two
+     passes are judged to be too costly.
+2. Make sure that consuming libraries are prepared for the breakage.
+   - In the case of in-place breakage, have pull requests (PRs) submitted to the
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: "have pull requests (PRs) submitted" --> submit pull requests

File: library-best-practices/JLBP-17.md [hunk 37]
```diff
@@ -0,0 +1,38 @@
...
+     consuming libraries that switch from the old surface to the new surface
+     (marked as DO NOT SUBMIT).
+     These PRs of course will not pass CI, but the author should at
+     least verify that the code builds locally before requesting a review.
+     The author should make sure that any code-level concerns of the library
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: of the --> the

File: library-best-practices/JLBP-17.md (hunk 34 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: I will rework this in a slightly different way. (Hopefully I address your implicit concerns)

File: library-best-practices/JLBP-17.md (hunk 35 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: That makes sense for the case of making something `final`, but for the checked exception case, you could technically introduce a new method whose difference is the exceptions it declares to throw. It's worth mentioning this type of change, though.

File: library-best-practices/JLBP-17.md (hunk 36 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Ok.

File: library-best-practices/JLBP-17.md (hunk 37 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: that's ungrammatical

**Comment by garrettjonesgoogle** [comment]
Another iteration - PTAL @elharo

File: library-best-practices/JLBP-17.md [hunk 38]
```diff
@@ -0,0 +1,44 @@
...
+   - If the feature is used by stable code in other libraries, you must use the
+     two-phase approach, except for the types of changes where this is
+     impossible (see below).
+   - If the code does not appear to be used in other libraries, an in-place
+     breakage may be ok.
+   - If the feature is used by unstable code or stable code having little usage
+     itself, prefer two phases if possible, with in-place breakage only if a
+     two-phase rollout is judged to be too costly.
+   - Some types of changes cannot be done using a two-phase approach, for
+     example, making classes or methods `final` or adding non-defaulted methods
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: non-defaulted methods --> methods to interfaces without a default implementation

File: library-best-practices/JLBP-17.md [hunk 39]
```diff
@@ -0,0 +1,44 @@
...
+     into interfaces. Minimize these types of changes.
+2. Make sure that consuming libraries are prepared for the breakage.
```
Comment (elharo) [inline library-best-practices/JLBP-17.md]: I'm still not convinced #2 is a good idea.

File: library-best-practices/JLBP-17.md (hunk 38 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: Will do.

File: library-best-practices/JLBP-17.md (hunk 39 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-17.md]: I established this as a process in my team and used it for about 2 years to manage breaking changes between gax-java and google-cloud-java when they were not yet fully stable, and it worked well because it prevented scenarios where people might check out both repos and find they couldn't build against HEAD of both. My recommendation here is out of experience.
